aim/sdk/adapters/optuna.py

```diff
@@ -0,0 +1,192 @@
+
+    def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -> None:
+
+        if isinstance(self._metric_name, str):
+            if len(trial.values) > 1:
+                # Broadcast default name for multi-objective optimization.
+                names = ["{}_{}".format(self._metric_name, i) for i in range(len(trial.values))]
+
+            else:
+                names = [self._metric_name]
+
+        else:
+            if len(self._metric_name) != len(trial.values):
+                raise ValueError(
+                    "Running multi-objective optimization "
+                    "with {} objective values, but {} names specified. "
+                    "Match objective values and names, or use default broadcasting.".format(
+                        len(trial.values), len(self._metric_name)
+                    )
+                )
+
+            else:
+                names = [*self._metric_name]
+
+        metrics = {name: value for name, value in zip(names, trial.values)}
+
+        if self._as_multirun:
+            metrics["trial_number"] = trial.number
+
+        attributes = {"direction": [d.name for d in study.directions]}
+
+        step = trial.number if self._run else None
+
+        if not self._run:
+            self.setup()
+            self._run["run_name"] = f"trial-{trial.number}"
+
+        for key, value in attributes.items():
+            self._run.set(('hparams', key), value, strict=False)
+
+        self._run.set("study_name", study.study_name)
+
+        if self._as_multirun:
+            for key, value in trial.params.items():
+                self._run.set(('hparams', key), value, strict=False)
+
+            for key, value in metrics.items():
+                self._run.set(key, value, strict=False)
+
+            self.close()
+        else:
+            for key, value in trial.params.items():
+                self._run.track(value, name=key, step=step)
+
+            for key, value in metrics.items():
+                self._run.track(value, name=key, step=step)
+
+    @property
+    def experiment(self) -> Run:
+        if self._run is not None:
+            return self._run
+
+    def setup(self, args=None):
+        if not self._run:
+            if self._run_hash:
+                self._run = Run(
+                    self._run_hash,
+                    repo=self._repo_path,
+                    system_tracking_interval=self._system_tracking_interval,
+                )
+            else:
+                self._run = Run(
+                    repo=self._repo_path,
+                    experiment=self._experiment_name,
+                    system_tracking_interval=self._system_tracking_interval,
+                    log_system_params=self._log_system_params,
+                )
+                self._run_hash = self._run.hash
+
+        if args:
+            for key, value in args.items():
+                self._run.set(key, value, strict=False)
+
+    def close(self) -> None:
```

Comment:
let's add `self._run` existence check, as this can be called by the user even before `self._run` is initialized